android/src/main/java/com/doordeck/headlessreactnativesdk/HeadlessReactNativeSdkModule.kt

```diff
@@ -0,0 +1,176 @@
+
+@ReactModule(name = HeadlessReactNativeSdkModule.NAME)
+class HeadlessReactNativeSdkModule(
+  reactContext: ReactApplicationContext,
+  private val doordeckSdk: Doordeck,
+) : NativeHeadlessReactNativeSdkSpec(reactContext) {
+
+  override fun getName(): String {
+    return NAME
+  }
+
+  /**
+   * Auth operations
+   */
+
+  override fun login(
+    email: String,
+    password: String,
+    promise: Promise,
+  ) {
+    setKeyPairIfNeeded()
+
+    doordeckSdk.accountless().loginAsync(email, password)
+      .thenApply {
+        respondNeedsVerification(promise)
+      }
+      .exceptionally { error ->
+        promise.reject("LOGIN_ERROR", error)
+      }
+  }
+
+  override fun setAuthToken(
+    authToken: String,
+    promise: Promise,
+  ) {
+    setKeyPairIfNeeded()
+
+    doordeckSdk.contextManager().setCloudAuthToken(authToken)
+    respondNeedsVerification(promise)
+  }
+
+  override fun verify(code: String, promise: Promise) {
+    doordeckSdk.account().verifyEphemeralKeyRegistrationAsync(code)
+      .thenApply {
+        promise.resolve(null)
+      }
+      .exceptionally { error ->
+        promise.reject("VERIFY_ERROR", error)
+      }
+  }
+
+  override fun logout(promise: Promise) {
+    doordeckSdk.account().logoutAsync()
+      .thenApply {
+        promise.resolve(null)
+      }
+      .exceptionally { error ->
+        promise.reject("LOGOUT_ERROR", error)
+      }
+  }
+
+  /**
+   * User operations
+   */
+  override fun getUserDetails(promise: Promise) {
+    doordeckSdk.account().getUserDetailsAsync()
+      .thenApply { response ->
+        promise.resolve(response.toNativeMap(
+          userId = doordeckSdk.contextManager().getUserId(),
+          certificateChainAboutToExpire = doordeckSdk.contextManager().isCertificateChainAboutToExpire(),
+          tokenAboutToExpire = doordeckSdk.contextManager().isCloudAuthTokenAboutToExpire(),
+        ))
+      }
+      .exceptionally {  error ->
+        promise.reject("USER_DETAILS_ERROR", error)
+      }
+  }
+
+  /**
+   * Device and tile operations
+   */
+
+  override fun getLocksBelongingToTile(tileId: String, promise: Promise) {
+    doordeckSdk.tiles().getLocksBelongingToTileAsync(tileId)
+      .thenApply { response ->
+        promise.resolve(response.toNativeMap())
+      }
+      .exceptionally { error ->
+        promise.reject("LOCKS_ERROR", error)
+      }
+  }
+
+  override fun unlockDevice(lockId: String, promise: Promise) {
+    doordeckSdk.lockOperations().unlockAsync(
+      LockOperations.UnlockOperation(
+        LockOperations.BaseOperation(
+          lockId = lockId,
+        )
+      )
+    )
+      .thenApply {
+        promise.resolve(null)
+      }
+      .exceptionally { error ->
+        promise.reject("UNLOCK_ERROR", error)
+      }
+  }
+
+
+  companion object {
+    const val NAME = "HeadlessReactNativeSdk"
+  }
+
+  private fun setKeyPairIfNeeded() {
+    if (doordeckSdk.contextManager().getKeyPair() == null) {
```

Comment:
Here you can use `isKeyPairValid()` instead, it also checks if the key pair is valid.

Reply:
your suggestion is better and I'm changing Swift as well